### ISSUE TITLE
Support of keyboard layouts list based on lastly used (3+ layouts)

### ISSUE THREAD

**Issue by Andrew15-5** [ankostis/gnome-shell-quick-lang-switch#7]
Is it possible to change (or add) the behavior to mimic the default GNOME layout switch?

Right now, this extension just iterates over all layouts in a loop (which have constant order).

But I would like to switch between 2 most recent layouts if the shortcut (Super+Space) is pressed once. And iterate over layouts if Super key is pressed down and Space key is pressing n times (for changing to nth layout from current in circle layout list).

I have layouts: a, b, c. and the lastly used order is the same with a being the current one.

- if I press `Super+Space` n times, then it will change between layout a & b n times
- if I hold down `Super` and press `Space`:
 1. time: current layout is b
 2. times: current layout is c
 3. times: current layout is a
 4. times: current layout is b
 5. times: current layout is c

And the tip of having a one stroke shortcut for forward/backward iteration is super useful, but I don't have any extra key to spare (especially which also would be convenient to press).

**Comment by ankostis** [comment]
I found [this script that utilizes `getInputSourceManager()._mruSources[1]`](https://gist.github.com/Envek/85f40478d1c8b9658621190569046447)
but it would only cycle between the last 2 input-sources.
I can't think of a way to utilize it without replicating `_modifiersSwitcher` logic.

**Comment by Andrew15-5** [comment]
The whole problem boils down to add ability to listen to `Super` key while it (and only it) stays pressed down (and count the number of times the space bar key has been pressed).

I'm only familiar with some (JS) methods that are used to get/set input sources because I made myself a little extension that provides some otherwise closed API for controlling input stuff (since GNOME 41 or higher, I can't remember). So I can't really help with this.

**Comment by osamuaoki** [comment]
FYI: I recently created GNOME shell extension to address this tedious input method switching issue. Instead using IM usage history which also can become confusing, I decided to bind one shortcut for each IM.  I also made these shortcuts to be configurable by the user via prefs.js.

- https://extensions.gnome.org/extension/6066/shortcuts-to-activate-input-methods/

I am using this under GNOME 43

**Comment by Andrew15-5** [comment]
> This is essential process to get GNOME shell extension to function as expected.

Currently, I don't have time to re-login. But it works fine for me. But I have 2 issues with "us (xkb)", "ru (xkb)" and "mozc-jp (ibus)":
- small but noticeable delay between pressing shortcut and able to type anything (0.5–1 s) when "ru" is involved,
- I have to go through "us" to be able to switch from "ru" to "jp".

I don't know if I have to open an issue [there](https://github.com/osamuaoki/inputmethod-shortcuts), because this problem was hunting me from the day 1 when I started using GNOME with en+ru keyboard layouts. I think this is some kind of bug around Russian layout (and of course GNOME), because without it everything works great. Since KDE doesn't have issue with slow layout switch (unlike GNOME), I think Russian layout won't be a problem either. But I have to switch to KDE one way or another to test this.

Not that I'm constantly changing from "ru" to "jp", but if I'm writing in "us", then I have to write in "ru". After a while I want to write in "jp", but I'm still writing in "ru (because of the bug)", which is very annoying (then I have to go back and forth to "us" and "ru").

**Comment by osamuaoki** [comment]
Ршб　... oops.

Hi,

I was wondering some odd situation with normal-us and intl-us to MOZC.  Now I see this problem clearer with Russian case.  Direct input in MOZC is Russian if I was from xkb-ru. 

Basically, All Japanese IM receive xkb output and uses it to convert either as Romaji or KANA conversion-table.  Anthy is configurable while MOZC is fixed.  So theoretically, you can create custom Romaji or KANA conversion-t able to move from ru to anthy.  But if you do that, en to Anthy becomes broken.  So this is not right approach.

I think generic fix for ibus input method is to set xkb=us first before setting ibus.

This is a bit complicated but interesting feature enhancement.  Please open ticket [there](https://github.com/osamuaoki/inputmethod-shortcuts).  I will consider this to be good mid-term project.

Basically, if there is more than 2 xkb settings, switching to ibus based input methods requires to go through "us" (or any default one such as "jp" with extra keys).  I probably make it pick the first xkb found.  There need good thought how this UI should be designed.

**Comment by osamuaoki** [comment]
I think newer version address this issue.

**Comment by ankostis** [comment]
I don't plan to provide this feature, due to the complexity of the implementation.

Either use [osamuaoki's extension _"Input method and touchpad shortcuts"_](https://extensions.gnome.org/extension/6066/shortcuts-to-activate-input-methods/), or someone else may provide a merge request that preserves both behaviors (with some config switch).

**Comment by ankostis** [comment]
FUI [v8](https://github.com/ankostis/gnome-shell-quick-lang-switch/releases/tag/v8) supports cycling "back" with [Shift]+... so a workaround is to arrange your layouts appropriately to cluster your most frequent ones close together so as to navigate back-n-forth.

Note that I'm not that eager to implement FIXED+MRU cycling bc i'm using a single key to cycle, eg **[PrtScn]** alone, which allows me to switch even faster - but the downside is that there is no "hold-down" state.

Still it can be done to cycle MRU layouts when pressing [Shift]+... (which now binds to backward cycling) and FIXED otherwise, or reversed.
That would employ configuration settings to select the preferred behavior, 
something that would increase the development effort.
